samples/apps/logging/logging.cpp

```diff
@@ -581,6 +582,140 @@ namespace loggingapp
+          view_of_final_seqno,
+          in.to_seqno,
+          view_of_final_seqno,
+          committed_view,
+          committed_seqno);
+        if (tx_status != ccf::TxStatus::Committed)
+        {
+          args.rpc_ctx->set_response_status(HTTP_STATUS_BAD_REQUEST);
+          args.rpc_ctx->set_response_header(
+            http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+          args.rpc_ctx->set_response_body(fmt::format(
+            "Only committed transactions can be queried. Transaction {}.{} is "
+            "{}",
+            view_of_final_seqno,
+            in.to_seqno,
+            ccf::tx_status_to_str(tx_status)));
+          return;
+        }
+
+        // Use hash of request as RequestHandle. WARNING: This means identical
+        // requests from different users will collide, and overwrite each
+        // other's progress!
+        ccf::historical::RequestHandle handle;
+        {
+          auto size =
+            sizeof(in.from_seqno) + sizeof(in.to_seqno) + sizeof(in.id);
+          std::vector<uint8_t> v(size);
+          auto data = v.data();
+          serialized::write(data, size, in.from_seqno);
+          serialized::write(data, size, in.to_seqno);
+          serialized::write(data, size, in.id);
+          handle = std::hash<decltype(v)>()(v);
+        }
+
+        // Fetch the requested range
+        auto& historical_cache = context.get_historical_state();
+
+        // TODO: Add pagination limit
+
+        auto stores =
+          historical_cache.get_store_range(handle, in.from_seqno, in.to_seqno);
+        if (stores.empty())
+        {
+          args.rpc_ctx->set_response_status(HTTP_STATUS_ACCEPTED);
+          static constexpr size_t retry_after_seconds = 3;
+          args.rpc_ctx->set_response_header(
+            http::headers::RETRY_AFTER, retry_after_seconds);
+          args.rpc_ctx->set_response_header(
+            http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+          args.rpc_ctx->set_response_body(fmt::format(
+            "Historical transactions from {} to {} are not yet "
+            "available, fetching now",
+            in.from_seqno,
+            in.to_seqno));
+          return;
+        }
+
+        // Process the fetched Stores
+        LoggingGetHistoricalRange::Out response;
+        for (size_t i = 0; i < stores.size(); ++i)
+        {
+          const auto store_seqno = in.from_seqno + i;
+          auto& store = stores[i];
+
+          auto historical_tx = store->create_read_only_tx();
+          auto records_handle = historical_tx.ro(records);
+          const auto v = records_handle->get(in.id);
+
+          if (v.has_value())
+          {
+            LoggingGetHistoricalRange::Entry e;
+            e.seqno = store_seqno;
+            e.id = in.id;
+            e.msg = v.value();
+            response.entries.push_back(e);
+          }
+          // We do not include an entry when the given key wasn't modified at
```

Comment:
Does that mean I cannot access other store data at that seqno if it hasn't been modified? Will this be possible in the future?

Reply:
Yes - this is how the historical query system currently operates. The retrieved store is constructed from the deserialised write set at that entry, so it only contains the state _written by_ that seqno. It does not contain the complete state that was visible at that seqno. The latter is possible to build with snapshots, but isn't needed so we haven't built it yet.